include/triton/Analysis/Utility.h

```diff
@@ -372,6 +372,53 @@ std::unique_ptr<DataFlowSolver> createDataFlowSolver();
 
 triton::MakeTensorPtrOp getMakeTensorPtrOp(Value v);
 
+namespace triton {
+
+template <typename T>
+SmallVector<T> applyPermutation(ArrayRef<T> vec,
+                                ArrayRef<int32_t> permutation) {
+  assert(vec.size() == permutation.size());
+
+  // Check that `permutation` is actually a permutation.
+#ifndef NDEBUG
+  SmallVector<int32_t> sortedPerm(permutation);
+  llvm::sort(sortedPerm);
+  for (int i = 0; i < sortedPerm.size(); ++i) {
+    assert(sortedPerm[i] == i);
+  }
+#endif
+
+  SmallVector<T> result;
+  result.reserve(vec.size());
+  for (auto i : permutation) {
+    result.push_back(vec[i]);
+  }
+  return result;
+}
+
+// These overloads are necessary to get applyPermutation() to work without an
+// explicit cast of the operands to ArrayRef.
+inline SmallVector<int32_t> applyPermutation(ArrayRef<int32_t> vec,
```

Comment:
You do not want them to be static.  Just `inline` is correct.

`inline` does not have to do with inlining the function in the optimizer.  Instead, it means that duplicate definitions of the function are allowed, and the linker is allowed to assume all definitions are the same (so it can pick one, under the ODR).

If we mark it as `static`, then this is saying that each definition of this function is unique -- in other words, each translation unit will get its own copy of the function.  It's not an ODR violation because they're each effectively in a unique anon ns.

I cannot think of a situation where it's correct to use `static` on a function or variable inside a header file.

Reply:
I see, I guess I didn’t realize it means weak linking. I thought I remembered it differently. Anyway..

Reply:
FWIW weak linkage means something different, corresponding to this idea  https://en.wikipedia.org/wiki/Weak_symbol :)

In LLVM `inline` corresponds to linkonce_odr linkage, https://llvm.org/docs/LangRef.html#linkage-types

Reply:
TIL :)